### ISSUE TITLE
"Edit Profile" bio must have a minimum of 30 characters.

### ISSUE THREAD

**Issue by ordnl** [codidact/qpixel#1092]
From https://github.com/codidact/qpixel/issues/945: when this is fixed, update these meta posts:
https://meta.codidact.com/posts/287128
https://meta.codidact.com/posts/287130

**Describe the bug**
Some markdown editors on the site have a 30 character minimum. This may not be desirable in some cases.

**To Reproduce**
1. Go to: /users/edit/profile
2. Edit "Profile" bio text to have < 30 characters.
3. Save button is disabled.

**Expected behavior**
I was expecting either:
* A helpful error message stating the 30 character minimum or
* To be able to have a bio with < 30 characters.

Instead the "Save" button was disabled silently.

**Environment**
 - OS: macOS Monterey
    - Browser: Chromium
    - Version: 103.0.5060.53
 - OS: Slackware Linux 15
    - Browser: Firefox
    - Version: 102.12.0esr
 - OS: Windows 10
    - Browser: Chrome
    - Version: 114.0.5735.198

**Additional context**
While I have no experience with Ruby On Rails, it seems like this is caused by the following:

* [Line 40: /app/views/users/edit_profile.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/users/edit_profile.html.erb#L40) uses `shared/body_field`  
* [Line 25: /app/views/shared/_body_field.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/shared/_body_field.html.erb#L25) uses `posts/mdhint`  
* [Line 5: /app/views/posts/_mdhint.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/posts/_mdhint.html.erb#L5) specifies `data-min="30"` which is enforced by:
  * [Line 43-49: /app/assets/javascripts/character_count.js](https://github.com/codidact/qpixel/blob/develop/app/assets/javascripts/character_count.js#L43-L49)

This means, changing `data-min` via Inspect Element will bypass this limit for "Edit Profile" since it is not enforced server side ([however it seems like it is for posts?](https://github.com/codidact/qpixel/blob/develop/db/schema.rb#L113))

Changing `data-min="30"` in `mdhint` will affect:
* [/app/views/tour/question2.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/tour/question2.html.erb#L33)
* [/app/views/admin/all_email.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/admin/all_email.html.erb#L12)
* [/app/views/mod_warning/new.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/mod_warning/new.html.erb#L36)
* [/app/views/tags/_form.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/tags/_form.html.erb#L70)
* [/app/views/posts/_form.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/posts/_form.html.erb#L57)

Perhaps, `body_field` & `mdhint` could accept `min` & `max` with default values set at `30`?

**Comment by trichoplax** [comment]
Related Meta post:
- [ Is there a reason for preventing user profile content being less than 30 characters? ](https://meta.codidact.com/posts/287130)

There is no answer yet, but a comment suggests there is no need to have a minimum character count for the user profile.

**Comment by ordnl** [comment]
The minimum was imposed in the following commit:
* [Update _mdhint.html.erb](https://github.com/codidact/qpixel/commit/7cc0f2cf8dad5135f1fb9673c241c1f1b24b1aeb)

It was in response to [a review request](https://github.com/codidact/qpixel/pull/608#pullrequestreview-724907740) in #608 

As suggested in the related meta post, this pull request was for post title & body, and inherited appropriately to:
* [/app/views/tour/question2.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/tour/question2.html.erb#L33)
* [/app/views/posts/_form.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/posts/_form.html.erb#L57)

And perhaps unintentionally to:
* [/app/views/admin/all_email.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/admin/all_email.html.erb#L12)
* [/app/views/mod_warning/new.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/mod_warning/new.html.erb#L36)
* [/app/views/tags/_form.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/tags/_form.html.erb#L70)
* [/app/views/users/edit_profile.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/users/edit_profile.html.erb#L40)

**Comment by cellio** [comment]
Profiles are empty by default (the placeholder text isn't in the profile), so we should allow you to revert to that state by deleting everything if you want to.

**Comment by trichoplax** [comment]
Related:
- #945 

Should that one be closed as a duplicate of this one, since this one has more detail?

**Comment by trichoplax** [comment]
It's perfectly possible to implement either of the following for the profile text:
- Have a 30 character minimum, but allow zero characters as an exception
- Have no minimum

Which is preferred? Note that the near-duplicate issue linked above is tagged "complexity:easy" and "meta:good-first-issue", so the decision on which approach to take it the only thing holding up someone implementing this.

My preference would be to have no minimum - I don't see any need for one.

**Comment by ordnl** [comment]
I would prefer no minimum on profile, tags, sending all emails and mod warnings. And a 30 character minimum on posts.

I have no experience in Rails and none of this code is tested, I do not know what the preferred code style for this project is.

However, I am thinking of something like this in [/app/views/posts/_mdhint.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/posts/_mdhint.html.erb):
```erb
 data-max="<%= max %>" data-min="<%= min %>" data-display-at="0.75">
 ```
And in [/app/views/shared/_body_field.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/shared/_body_field.html.erb)
 ```erb
<% min_body_length ||= 0 %>
<% max_body_length ||= 30000 %>
...
<%= render 'posts/mdhint',  min: min_body_length, max: max_body_length %>
 ```
The intent is to make `body_field` have a default minimum of 0 characters. With `post` and `tour` being the exceptions.


The exceptions can be made by changing [Line 33: /app/views/tour/question2.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/tour/question2.html.erb#L33) to:
```ruby
<%= render 'posts/mdhint', min: 30 %>
```
And [Line 57: /app/views/posts/_form.html.erb](https://github.com/codidact/qpixel/blob/develop/app/views/posts/_form.html.erb#L57) to:
```ruby
<%= render 'shared/body_field', f: f, field_name: :body_markdown, field_label: t('posts.body_label'), post: post, min_body_length: 30 %>
```

**Comment by cellio** [comment]
I don't see a need for a minimum.  It feels weird to allow 0, forbid 5, and allow 30.

Email and mod warnings should keep the minimum to avoid disruptive mistakes (those send email or notifications and should definitely have content).  Profiles and tag descriptions shouldn't have a minimum length.

**Comment by cellio** [comment]
https://github.com/codidact/qpixel/issues/945, which I've closed as a duplicate of this, has the following tags: 
- meta: good first issue
- meta: help wanted
- complexity: easy

I'm not sure which of those should transfer here, since I think we're discussing a different fix here?

**Comment by Taeir** [comment]
Fixed in #1156

**Comment by Taeir** [comment]
Todo: update meta posts linked
